OKTA-459416 Custom Self-Service Pwd Recovery Guide for Java (#3292)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/guides/oie-embedded-sdk-use-case-custom-pwd-recovery-mfa/main/java/siw.md

> > **Note**: The examples in this guide use Java 11 and Spring Boot MVC

Reviewer: ```suggestion
> **Note**: The examples in this guide use Java 11 and Spring Boot MVC.
```

Author: fixed

### 1: Start password recovery

> After the user starts the password recovery flow and selects the email authenticator for the process, Okta sends them an email matching the **Forgot Password** template you altered earlier.

Reviewer: ```suggestion
After the user starts the password recovery flow and selects the email authenticator for the process, Okta sends them an email that matches the **Forgot Password** template that you altered earlier.
```

Author: fixed

<div class="common-image-format">

![Example of email sent to user](/img/advanced-use-cases/custom-pwd-recovery-custom-email.png "Password recovery email")

</div>

> When the user clicks the **Reset Password** link, their browser sends a request to the endpoint defined by the template, attaching the `${oneTimePassword}` and `${request.relayState}` VTL variables as query parameters to the URL. For instance, in the sample this request might render as `http://localhost:8080/magic-link/callback?otp=726009&state=1b34371af02dd31d2bc4c48a3607cd32`.

Reviewer: ```suggestion
When the user clicks the **Reset Password** link, their browser sends a request to the endpoint defined by the template and attaches the `${oneTimePassword}` and `${request.relayState}` VTL variables as query parameters to the URL. For instance, in the sample this request might render as `http://localhost:8080/magic-link/callback?otp=726009&state=1b34371af02dd31d2bc4c48a3607cd32`.
```

Author: fixed

### 2: Handle the OTP and state parameters

Create a callback handler that takes the `${oneTimePassword}` and `${request.relayState}` values from the query string, saves them into local variables (for instance, 'otp', and 'state'), and makes the following checks:

> 1. That the current session attribute `IdxClientContext` is not `null` (because the user clicked the magic link from a different browser).

Reviewer: ```suggestion
1. The current session attribute `IdxClientContext` isn't `null` (because the user clicked the magic link from a different browser).
```

Author: Changed isn't, but didn't lose "That" for the same reason as above.

> 2. That `state` is not `null`.

Reviewer: ```suggestion
2. The `state` isn't `null`.
```

Author: Changed isn't, but didn't lose "That" for the same reason as above

> If either check returns false, redirect to an error page. If both checks return true, pass `IdxClientContext`,`otp` and `state` as session attributes to a page that contains the Sign-In Widget for processing. For example:

Reviewer: ```suggestion
If either check returns false, redirect the browser to an error page. If both checks return true, pass `IdxClientContext`,`otp` and `state` as session attributes to a page that contains the Sign-In Widget for processing. For example:
```

Author: fixed

Okta then tells the user to either click the link in the email or enter the code to continue and sends an email to their email address matching the Forgot Password template that was altered earlier.

<div class="common-image-format">

![Example of email sent to user](/img/advanced-use-cases/custom-pwd-recovery-custom-email.png "Password recovery email")

</div>

The email's **Reset Password** link includes the `otp` and `request.relayState` variables sent back as query parameters to the application. For instance, the URL in the email template,  `http://localhost:8080/magic-link/callback?otp=${oneTimePassword}&state=${request.relayState}`, might be rendered as `http://localhost:8080/magic-link/callback?otp=726009&state=1b34371af02dd31d2bc4c48a3607cd32` in the email sent to the user.

### 2: Handle the OTP and state parameters

Create a callback handler method that takes the `otp` and `state` parameters in the query string and passes them as session parameters to a page that contains the Sign-In Widget. Before calling the callback handler method, retrieve the current `IDXClientContext` object from the browser session and check that `state` has a value.

> If `state` is `null` or the browser context object is `null` (because the user clicked the magic link from a different browser), the widget instructs advise the user to return to the original tab in the browser where they requested a password reset and enter the OTP to proceed.

Reviewer 2: See my comment above for the same use of these parameters. The current text is a bit confusing and isn't in the appropriate sequence.

Author: Changed to

> Create a callback handler that takes the `otp` and `state` parameters from the query string, attempts to retrieve the current `IDXClientContext` object from the browser session, and then passes them all as session parameters to a page that contains the Sign-In Widget for processing.
> 
> If either the browser context object is `null` (because the user clicked the magic link from a different browser) or `state` is `null`, redirect to an error page instead.

Reviewer 2: This has the same issue as the previous one. Who calls the handler (and possible when)? What are the arguments for the handler (the current wording is unclear). The structure is also a bit misleading as it reads like: "createa a callback handler with no parameters that's called magically and extracts the otp and state parameters from the query string, then retrievs the ... etc."

BTY, "attempts to retrieve" is another "could" type expression. Either `IDXClientContext` is retrieved or it's not. If failure to retrieve it is an important error, then call it out.

Author: Changed to

> Create a mapping that maps a request for the magic link URL to a method that takes the `otp` and `state` parameters from the query string, and makes the following checks:
> 
> 1. That `otp` and `state` are not `null`.
> 2. That `state` matches the state stored in the current `ProceedContext` session variable.
> 
> If either check returns false, throw an error.

Idly, the concept of "attempting to do a thing" is baked into code isn't it? That's why they are called try..catch, and if...then clauses. Similarly here we have attempts to retrieve a value from several different locations and if the value isn't anywhere something else happens instead. I've changed it now in the text, but....

```java
@GetMapping("/magic-link/callback")
public ModelAndView handleMagicLinkCallback(
   HttpServletRequest request,
   @RequestParam(name = "state") String state,
   @RequestParam(name = "otp") String otp,
   HttpSession session) throws MalformedURLException
{
  if (session.getAttribute(IDX_CLIENT_CONTEXT) == null) {
      try {
         idxClientContext = idxAuthenticationWrapper.getClientContext();
      }
      catch (ProcessingException e) {
         // return error details
      }
      session.setAttribute(IDX_CLIENT_CONTEXT, idxClientContext);
  }

  if (idxClientContext == null) {
     // return error details
  }

  if (state == null) {
      return new ModelAndView("redirect:" + oktaOAuth2Properties.getRedirectUri());
  }

```

> If both values are valid, pass them (along with all the values required by the widget) to the page hosting the widget in your application.

Reviewer 2: "both values", which values?

> "If the values for ... are valid, pass them (and all the other required values) to the page in your app page that is hosting the widget.

It's also not clear that the parenthetical statement is necessary. It seems to be clear from context.

Author: Have deleted the whole text here and placed it at the top of the code block instead.

```java
  String issuer = oktaOAuth2Properties.getIssuer();
  // the widget needs the base url, just grab the root of the issuer
  String orgUrl = new URL(new URL(issuer), "/").toString();

  ModelAndView mav = new ModelAndView("login");
  mav.addObject(STATE, state);
  mav.addObject(OTP, otp);
  mav.addObject(SCOPES, oktaOAuth2Properties.getScopes());
  mav.addObject(OKTA_BASE_URL, orgUrl);
  mav.addObject(OKTA_CLIENT_ID, oktaOAuth2Properties.getClientId());
  mav.addObject(INTERACTION_HANDLE, idxClientContext.getInteractionHandle());
  mav.addObject(CODE_VERIFIER, idxClientContext.getCodeVerifier());
  mav.addObject(CODE_CHALLENGE, idxClientContext.getCodeChallenge());
  mav.addObject(CODE_CHALLENGE_METHOD, CODE_CHALLENGE_METHOD_VALUE);
  mav.addObject(REDIRECT_URI,
    request.getScheme() + "://" + request.getServerName() + ":" +
    request.getServerPort() + request.getContextPath() +
    "/authorization-code/callback");
  mav.addObject(ISSUER_URI, issuer);
  session.setAttribute(CODE_VERIFIER, idxClientContext.getCodeVerifier());
  return mav;
}
```

> You can pass the values to the widget using the widget's `config` object.

Reviewer 2: How about something simpler:

> Pass the values using the widget's `config` object.

I say this because "to the widget" is implied by using both "pass" and "using the widget's"

Author: Absolutely - I remember struggling to phrase this well.

```html
  <div id="sign-in-widget"></div>

  <script th:inline="javascript">
      /*<![CDATA[*/

      var config = {};
      config.baseUrl = /*[[${oktaBaseUrl}]]*/ 'https://{yourOktaDomain}';
      config.clientId = /*[[${oktaClientId}]]*/ '{clientId}';
      config.otp = /*[[${otp}]]*/ '{otp}';
      config.redirectUri = /*[[${redirectUri}]]*/ '{redirectUri}';
      config.useInteractionCodeFlow = true;
      config.interactionHandle = /*[[${interactionHandle}]]*/ '{interactionHandle}';
      config.codeChallenge = /*[[${codeChallenge}]]*/ '{codeChallenge}';
      config.codeChallengeMethod = /*[[${codeChallengeMethod}]]*/ '{codeChallengeMethod}';
      config.redirect = 'always';
      config.authParams = {
          issuer: /*[[${issuerUri}]]*/ '{issuerUri}',
          pkce: true,
          state: /*[[${state}]]*/ '{state}' || false,
          nonce: /*[[${nonce}]]*/ '{nonce}',
          scopes: /*[[${scopes}]]*/ '[scopes]',
      };

      new OktaSignIn(config).showSignInAndRedirect(
          { el: '#sign-in-widget' },
          function (res) {}
      );
      /*]]>*/
  </script>
```

### 3: Displays password reset prompt and complete password recovery flow

> When the widget loads, it checks whether the `state` and `otp` values are valid with the Okta server. If the values are valid, the following reset password page appears for the user to enter their new password. The user continues the password recovery flow described in the [User password recovery summary of steps](/docs/guides/oie-embedded-sdk-use-case-pwd-recovery-mfa/java/main/#summary-of-steps).

Reviewer 2: The widget loads and then checks the validity of `state` and `otp` with the Okta server. If the values are valid, the user is prompted to reset their password. (what "following page", I don't see one... or is that referring to a new page that's loaded by the or in the widget?)

Author: Have changed the text - there is a screenshot of the widget's password reset form shown directly beneath this text.

### 1: Start password recovery

To begin the password recovery flow, the user must:

> 1. Click the **Forgot Password?** link on the sign-in page.

Reviewer 3: ```suggestion
1. Clicks the **Forgot Password?** link on the sign-in page.
```

Author: fixed

> 2. Enter their **Email or Username** in the box and click **Next**.

Reviewer 3: ```suggestion
2. Enters their **Email or Username** in the password recovery dialog and clicks **Next**.
```

> 3. Choose **Email** as the authenticator they want to use for password recovery and click **Submit**.

Reviewer 3: ```suggestion
3. Chooses **Email** as the authenticator they want to use for password recovery and clicks **Submit**.
```

Author: fixed

> Okta then sends the user an email that matches the Forgot Password template that you altered earlier, and the Sign-In Widget tells them to click the link in the email or to enter the OTP to continue.

Reviewer 3: ```suggestion
Okta then sends the user an email that matches the Forgot Password template that you altered earlier. The Sign-In Widget instructs the user to click the link in the email or to enter the OTP to continue.
```

Author: fixed

<div class="common-image-format">

> ![Screenshot of email sent to user](/img/advanced-use-cases/custom-pwd-recovery-custom-email.png "Password Recovery Email")

Reviewer 3: ```suggestion
![Example of email sent to user](/img/advanced-use-cases/custom-pwd-recovery-custom-email.png "Password recovery email")
```
Maybe we don't need this image since it was already displayed for SDK steps?

Author: fixed. It's in the other guides so I'll leave this as is.

</div>

> The email's **Reset Password** link includes the `otp` and `request.relayState` variables sent back as query parameters to the application. For example,

Reviewer 3: ```suggestion
The email's **Reset Password** link includes the `otp` and `request.relayState` query parameters.
```

> `http://localhost:8080/magic-link/callback?otp=${oneTimePassword}&state=${request.relayState}` becomes `http://localhost:8080/magic-link/callback?otp=726009&state=1b34371af02dd31d2bc4c48a3607cd32`.

Reviewer 3: ```suggestion
For example, the email template URL, `http://localhost:8080/magic-link/callback?otp=${oneTimePassword}&state=${request.relayState}`, translates to the following URL request to your app:  `http://localhost:8080/magic-link/callback?otp=726009&state=1b34371af02dd31d2bc4c48a3607cd32`.
```

Author: fixed

### 2: Handle the OTP and state parameters

> Create a callback handler method that takes the `otp` and `state` parameters in the query string and passes them as session parameters to a page that contains the Sign-In Widget. First, retrieve the current `IDXClientContext` object from the browser session and check that `state` has a value.

Reviewer 3: ```suggestion
Create a callback handler method that takes the `otp` and `state` parameters in the query string and passes them as session parameters to a page that contains the Sign-In Widget. Before calling the callback handler method, retrieve the current `IDXClientContext` object from the browser session and check that `state` has a value.
```

Author: fixed

> If `state` is `null` or the browser context object is `null` (because the user is clicking the magic link in a different browser), the widget will advise the user to return to the original tab in the browser where they requested a password reset and enter the OTP to proceed.

Reviewer 3: ```suggestion
If `state` is `null` or the browser context object is `null` (because the user clicked the magic link from a different browser), the widget instructs the user to return to the original tab in the browser where they requested a password reset and enter the OTP to proceed.
```

Author: fixed

```java
@GetMapping("/magic-link/callback")
public ModelAndView handleMagicLinkCallback(
   HttpServletRequest request,
   @RequestParam(name = "state") String state,
   @RequestParam(name = "otp") String otp,
   HttpSession session) throws MalformedURLException
{
  if (session.getAttribute(IDX_CLIENT_CONTEXT) == null) {
      try {
         idxClientContext = idxAuthenticationWrapper.getClientContext();
      }
      catch (ProcessingException e) {
         // return error details
      }
      session.setAttribute(IDX_CLIENT_CONTEXT, idxClientContext);
  }

  if (idxClientContext == null) {
     // return error details
  }

  if (state == null) {
      return new ModelAndView("redirect:" + oktaOAuth2Properties.getRedirectUri());
  }

```

> If both values are valid, collate all the values required by the widget and pass them to the page hosting the widget in your application.

Reviewer 3: ```suggestion
If both values are valid, pass them (along with all the values required by the widget) to the page hosting the widget in your application.
```

Author: fixed

```html
  <div id="sign-in-widget"></div>

  <script th:inline="javascript">
      /*<![CDATA[*/

      var config = {};
      config.baseUrl = /*[[${oktaBaseUrl}]]*/ 'https://{yourOktaDomain}';
      config.clientId = /*[[${oktaClientId}]]*/ '{clientId}';
      config.otp = /*[[${otp}]]*/ '{otp}';
      config.redirectUri = /*[[${redirectUri}]]*/ '{redirectUri}';
      config.useInteractionCodeFlow = true;
      config.interactionHandle = /*[[${interactionHandle}]]*/ '{interactionHandle}';
      config.codeChallenge = /*[[${codeChallenge}]]*/ '{codeChallenge}';
      config.codeChallengeMethod = /*[[${codeChallengeMethod}]]*/ '{codeChallengeMethod}';
      config.redirect = 'always';
      config.authParams = {
          issuer: /*[[${issuerUri}]]*/ '{issuerUri}',
          pkce: true,
          state: /*[[${state}]]*/ '{state}' || false,
          nonce: /*[[${nonce}]]*/ '{nonce}',
          scopes: /*[[${scopes}]]*/ '[scopes]',
      };

      new OktaSignIn(config).showSignInAndRedirect(
          { el: '#sign-in-widget' },
          function (res) {}
      );
      /*]]>*/
  </script>
```

### 3: Displays password reset prompt and complete password recovery flow

> After the widget is loaded, it checks whether `state` and `otp` are valid with the Okta server. Assuming they are, either the following reset page or a prompt appears for the user to enter the OTP. After the user enters the OTP, the reset page appears. The user continues the password recovery flow described in the [User password recovery guide](/docs/guides/oie-embedded-sdk-use-case-pwd-recovery-mfa/java/main/).

Reviewer 3: ```suggestion
When the widget loads, it checks whether the `state` and `otp` values are valid with the Okta server. If the values are valid, the following reset password page appears for the user to enter their new password. The user continues the password recovery flow described in the [User password recovery summary of steps](/docs/guides/oie-embedded-sdk-use-case-pwd-recovery-mfa/java/main/#summary-of-steps).
```
Do we need to tell the user to enter the OTP if the `otp` value passed to Okta is valid?

Author: no

<div class="common-image-format">

> ![Screenshot of password reset page](/img/advanced-use-cases/java-custom-pwd-recovery-custom-siw-reset-pwd-page.png "Password Reset Page")

Reviewer 3: This image is quite big compared to the sequence diagrams. Is there any way to reduce the size? Or maybe not display this image if it's not required?

### 1: Start password recovery

> To begin the password recovery flow, the user must

Author: fixed

1. Click the **Forgot Password?** link on the sign-in page.
2. Enter their **Email or Username** in the box and click **Next**.
3. Choose **Email** as the authenticator they want to use for password recovery and click **Submit**.

> Okta then sends the user an email that matches the Forgot Password template that was altered earlier and the widget tells them either to click the link in the email or enter the OTP to continue.

Author: fixed

<div class="common-image-format">

![Screenshot of email sent to user](/img/advanced-use-cases/custom-pwd-recovery-custom-email.png "Password Recovery Email")

</div>

> The **Reset Password** link in the email includes the `otp` and `request.relayState` variables as query parameters back to the application. For example,

Author: fixed

If `state` is `null` or the browser context object is `null` (because the user is clicking the magic link in a different browser), the widget will advise the user to return to the original tab in the browser where they requested a password reset and enter the OTP to proceed.

```java
@GetMapping("/magic-link/callback")
public ModelAndView handleMagicLinkCallback(
   HttpServletRequest request,
   @RequestParam(name = "state") String state,
   @RequestParam(name = "otp") String otp,
   HttpSession session) throws MalformedURLException
{
  if (session.getAttribute(IDX_CLIENT_CONTEXT) == null) {
      try {
         idxClientContext = idxAuthenticationWrapper.getClientContext();
      }
      catch (ProcessingException e) {
         // return error details
      }
      session.setAttribute(IDX_CLIENT_CONTEXT, idxClientContext);
  }

  if (idxClientContext == null) {
     // return error details
  }

  if (state == null) {
      return new ModelAndView("redirect:" + oktaOAuth2Properties.getRedirectUri());
  }

```

> If both values are valid, collate all the values required by the widget and pass them to the default Redirect URI for your app integration.

Reviewer 4: Can we mention how setting these values here relates to the following javascript just so folks can understand how the OTP and other values pass to the widget: 

      /*<![CDATA[*/

      var config = {};

      config.baseUrl = /*[[${oktaBaseUrl}]]*/ 'https://{yourOktaDomain}';
      config.clientId = /*[[${oktaClientId}]]*/ '{clientId}';
      config.otp = /*[[${otp}]]*/ '{otp}';
      config.redirectUri = /*[[${redirectUri}]]*/ '{redirectUri}';
      config.useInteractionCodeFlow = true;
      config.interactionHandle = /*[[${interactionHandle}]]*/ '{interactionHandle}';
      config.codeChallenge = /*[[${codeChallenge}]]*/ '{codeChallenge}';
      config.codeChallengeMethod = /*[[${codeChallengeMethod}]]*/ '{codeChallengeMethod}';
      config.redirect = 'always';
      config.authParams = {
          issuer: /*[[${issuerUri}]]*/ '{issuerUri}',
          pkce: true,
          state: /*[[${state}]]*/ '{state}' || false,
          nonce: /*[[${nonce}]]*/ '{nonce}',
          scopes: /*[[${scopes}]]*/ '[scopes]',
      };

      new OktaSignIn(config).showSignInAndRedirect(
          { el: '#sign-in-widget' },
          function (res) {}
      );
      /*]]>*/

Author: I've added a note quoting this code and how it passes the values into the widget.

```java
  String issuer = oktaOAuth2Properties.getIssuer();
  // the widget needs the base url, just grab the root of the issuer
  String orgUrl = new URL(new URL(issuer), "/").toString();

  ModelAndView mav = new ModelAndView("login");
  mav.addObject(STATE, state);
  mav.addObject(OTP, otp);
  mav.addObject(SCOPES, oktaOAuth2Properties.getScopes());
  mav.addObject(OKTA_BASE_URL, orgUrl);
  mav.addObject(OKTA_CLIENT_ID, oktaOAuth2Properties.getClientId());
  mav.addObject(INTERACTION_HANDLE, idxClientContext.getInteractionHandle());
  mav.addObject(CODE_VERIFIER, idxClientContext.getCodeVerifier());
  mav.addObject(CODE_CHALLENGE, idxClientContext.getCodeChallenge());
  mav.addObject(CODE_CHALLENGE_METHOD, CODE_CHALLENGE_METHOD_VALUE);
  mav.addObject(REDIRECT_URI,
    request.getScheme() + "://" + request.getServerName() + ":" +
    request.getServerPort() + request.getContextPath() +
    "/authorization-code/callback");
  mav.addObject(ISSUER_URI, issuer);
  session.setAttribute(CODE_VERIFIER, idxClientContext.getCodeVerifier());
  return mav;
}
```

### 3: Displays password reset prompt and complete password recovery flow

> After the widget is loaded, it checks whether `state` and `otp` are valid with the Okta server. Assuming they are, either the following reset page appears, or a prompt appears for the user to enter the OTP. After the user enters the OTP, the reset page appears. The user continues the password recovery flow described in the [User password recovery guide](/docs/guides/oie-embedded-sdk-use-case-pwd-recovery-mfa/java/main/).

Author: fixed
